File: amazon/java/longest_substring_k_distinct.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        // if (s == null || s.length() == 0 || k == 0) return 0;
        HashMap<Character, Integer> d = new HashMap<Character, Integer>();
        int left = 0, result = 0;
        for (int i = 0; i < s.length(); i++) {
            d.put(s.charAt(i), i);
            if (d.size() > k) {
                int min = Collections.min(d.values());
                left = min + 1;
                d.remove(s.charAt(min));
            }
            result = Math.max(result, i - left + 1);
        }
        return result;
    }
}

// If required to return all the substrings (or the # of substrings) to have exactly k distinct letters
class Solution {
    public String[] lengthOfLongestSubstringKDistinct(String s, int k) {
        // if (s == null || s.length() == 0 || k == 0) return 0;
        HashMap<Character, Integer> d = new HashMap<Character, Integer>();
        List<String> l = new ArrayList<String>();
        int left = 0, result = 0;
        for (int i = 0; i < s.length(); i++) {
            d.put(s.charAt(i), i);
            if (d.size == k) {
                l.add(s.substring(left, i + 1));
            }
            if (d.size() > k) {
                int min = Collections.min(d.values());
                left = min + 1;
                d.remove(s.charAt(min));
            }
            result = Math.max(result, i - left + 1);
        }
        return l.toArray(new String[l.size()]);
    }
}

// find exactly k distinct letters
public class CountKSubStr 
{ 
    // Function to count number of substrings 
    // with exactly k unique characters 
    int countkDist(String str, int k) 
    { 
        // Initialize result 
        int res = 0; 
  
        int n = str.length(); 
  
        // To store count of characters from 'a' to 'z' 
        int cnt[] = new int[26]; 
  
        // Consider all substrings beginning with 
        // str[i] 
        for (int i = 0; i < n; i++) 
        { 
            int dist_count = 0; 
  
            // Initializing count array with 0 
            Arrays.fill(cnt, 0); 
  
            // Consider all substrings between str[i..j] 
            for (int j=i; j<n; j++) 
            { 
                // If this is a new character for this 
                // substring, increment dist_count. 
                if (cnt[str.charAt(j) - 'a'] == 0) 
                    dist_count++; 
  
                // Increment count of current character 
                cnt[str.charAt(j) - 'a']++; 
  
                // If distinct character count becomes k, 
                // then increment result. 
                if (dist_count == k) 
                    res++; 
            } 
        } 
  
        return res; 
    } 
  
    // Driver Program 
    public static void main(String[] args) 
    { 
        CountKSubStr ob = new CountKSubStr(); 
        String ch = "abcbaa"; 
        int k = 3; 
        System.out.println("Total substrings with exactly " + 
                           k +    " distinct characters : "
                           + ob.countkDist(ch, k)); 
    } 
}

File: amazon/python/longest_substring_k_distinct.py
def lengthOfLongestSubstringKDistinct(self, s, k):
    """
    :type s: str
    :type k: int
    :rtype: int
    """
    if not s or k == 0:
        return 0
    dic = collections.defaultdict(int)
    result = 0
    left = 0
    for i, letter in enumerate(s):
        dic[letter] += 1
        while len(dic) > k:
            dic[s[left]] -= 1
            if dic[s[left]] == 0:
                del dic[s[left]]
            left += 1
        result = max(result, i - left + 1)
    return result


# Or slightly better way:
def lengthOfLongestSubstringKDistinct(self, s, k):
    """
    :type s: str
    :type k: int
    :rtype: int
    """
    if not s or k == 0:
        return 0
    dic = {}
    left = 0
    result = 0
    for i, letter in enumerate(s):
        dic[letter] = i
        if len(dic) > k:
            to_del = min(dic.values())
            del dic[s[to_del]]
            left = to_del + 1
        result = max(result, i - left + 1)
    return result
# Find exactly k distinct letters


def countkDist(str1, k):
    n = len(str1)

    # Initialize result
    res = 0

    # To store count of characters from
    # 'a' to 'z'
    cnt = [0] * 27

    # Consider all substrings beginning
    # with str[i]
    for i in range(0, n):
        dist_count = 0

        # Initializing array with 0
        cnt = [0] * 27

        # Consider all substrings between str[i..j]
        for j in range(i, n):

            # If this is a new character for this
            # substring, increment dist_count.
            if(cnt[ord(str1[j]) - 97] == 0):
                dist_count += 1

            # Increment count of current character
            cnt[ord(str1[j]) - 97] += 1

            # If distinct character count becomes k,
            # then increment result.
            if(dist_count == k):
                res += 1
            if(dist_count > k):
                break

    return res
